### ISSUE TITLE
What's needed to include ClearStacktrace.jl in Base?

### ISSUE THREAD

**Issue by jkrumbiegel** [JuliaLang/julia#36026]
The title says it all. I think I figured out a good format to present stack traces, and I think all users could benefit from it. It's currently in [ClearStacktrace.jl](https://github.com/jkrumbiegel/ClearStacktrace.jl) and depends on `Crayons.jl` for the colors.

I want to know what I can do to make this PR-ready. I don't know how Julia Base handles REPL colors, which ones I'm allowed to use, how I access them without `Crayons` etc. I also don't know if there might be system-specific intricacies of stack traces that I have overlooked, testing this only a Windows and a Mac machine.

The basic idea is to have three columns, function, module and signature. Function and module together should basically always fit in a REPL horizontally, while signatures can be very long and are therefore allowed to overflow into new lines, while staying intact when resizing the REPL (compared to more complex but brittle table layouting). The code paths get a new line each, and are also allowed to overflow if they are too long. This keeps clickable links in Atom / VSCode and the like intact.

Just for reference, here is the before and after comparison from the README:
Before:
![grafik](https://user-images.githubusercontent.com/22495855/82832508-99862600-9ebb-11ea-90d3-fa0cb54eafe8.png)
After:
![grafik](https://user-images.githubusercontent.com/22495855/82832522-9f7c0700-9ebb-11ea-9076-6122ad482800.png)

**Comment by cossio** [comment]
This would be really nice to have. Or at least maybe a trimmed down version which improves printing of stacktraces by default in Base.

**Comment by JeffBezanson** [comment]
I'm all in favor. The main thing I don't get is the colors in the signatures. They seem...random? I guess the different colors are there to make it easier to pick out different elements, but it's a bit weird. How about color == nesting depth? I think the main weirdness is in e.g. `Tuple{Symbol,Symbol}` where the two `Symbol`s are different colors.

**Comment by timholy** [comment]
There's another thing I'd *really* like to see:
```julia
julia> foo(x::T, y::T) where T = error("whoops")
foo (generic function with 1 method)

julia> function bar()
           a = rand(3, 5)
           b = view(a, :, 2:3)
           c = reinterpret(Float32, b)
           foo(c, c)
       end
bar (generic function with 1 method)

julia> bar()
ERROR: whoops
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] foo(::Base.ReinterpretArray{Float32,2,Float64,SubArray{Float64,2,Array{Float64,2},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64}},true}}, ::Base.ReinterpretArray{Float32,2,Float64,SubArray{Float64,2,Array{Float64,2},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64}},true}}) at ./REPL[1]:1
 [3] bar() at ./REPL[2]:5
 [4] top-level scope at REPL[3]:1
```
but since
```julia
julia> m = first(methods(foo))
foo(x::T, y::T) where T in Main at REPL[1]:1

julia> m.sig
Tuple{typeof(foo),T,T} where T
```
it seems that we *should* be able to print entry 2 as something like
```
[2] foo(::T, ::T) where T = Base.ReinterpretArray{Float32,2,Float64,SubArray{Float64,2,Array{Float64,2},Tuple{Base.Slice{Base.OneTo{Int64}},UnitRange{Int64}},true}}
```

**Comment by KristofferC** [comment]
A small comment, but to me it is currently maybe a bit too much color "salad". The stackframe counter being blue, all the different colors in the signature etc. For Base colors so far, we've only used the 16 system colors (8 + 8 bright ones) and that makes fancy coloring a bit harder.

**Comment by jkrumbiegel** [comment]
Good suggestions overall. Maybe a couple comments why I made the color choices so far:

The numbers are blue because they are important, so shouldn't be grayed out, but making them white felt to me like they were fighting for attention with the function names.

The modules use the primary colors in a cycle, I think this is very useful because it allows to quickly filter the trace for relevant modules. This is something that is very hard right now. The idea is that people should be forced to read as little as possible.

The colors for the signatures are very debatable. The ones on the screenshot above were chosen to be only slightly different from dark gray. At them same time, I wanted the different semantic components to be discriminable. That’s why all types and type parameters cause a color switch. There are three colors which means there won’t be two neighbors with the same color. I found that it’s quite difficult to make out single words in huge type signatures if they are all just white.

As the 16 color set doesn’t have colors like the ones you see here it’s probably not something that can be included in Base. In a recent update, I replaced the colors with three slightly different shades of gray from the ANSI set, but even that will not be supported everywhere I guess.

**Comment by jkrumbiegel** [comment]
Oh also the stack frame color and the module colors as well as the dark gray and white are all system colors. It’s just my particular VSCode theme that renders them like you see here.

**Comment by mcabbott** [comment]
This would be great to have. 

When scanning argument types, often it's pretty hard to figure out where one stops and the next starts. Colouring the top level differently might be very helpful here, and perhaps even numbering the arguments:
<pre>
[3]   <b>(_1::DataFrames.var"#fun##309#"{</b>...lighter...<b>}, _2::NamedTuple{</b>...lighter...<b>}, _3::Int64})</b>
</pre>

**Comment by JeffBezanson** [comment]
We should be able to show the argument names, like we do in the printing of `methods(f)`.

**Comment by jkrumbiegel** [comment]
how's this? I've replaced the signature colors with dark gray, but the `::` is white so it's salient where argument types begin. especially helpful with the type monster at position 11 
<img width="604" alt="grafik" src="https://user-images.githubusercontent.com/22495855/82930580-e9312400-9f85-11ea-94e3-96db74ecb378.png">

**Comment by JeffBezanson** [comment]
I like it. Will be even better with argument names.

It's too bad that the order "function - module - arguments" works so well for the layout but doesn't really make sense semantically. I don't have a solution though; clearly it's best for the function to come first. Maybe the module could go on the next line before the file location? It is part of the location after all. Then maybe we wouldn't need the header either.

I think we could also put the numbers in normal text, and the function names in bold and/or white.

**Comment by timotaularson** [comment]
Could the filename at the end of the codepath (and possibly the line number) get their own bright color too?

**Comment by jkrumbiegel** [comment]
> Maybe the module could go on the next line before the file location? It is part of the location after all. Then maybe we wouldn't need the header either.

I'll try out a couple things with these suggestions.

> Could the filename at the end of the codepath (and possibly the line number) get their own lighter color too?

It technically can of course ;) It's always a tradeoff between better visibility and attention-grabbing noise.

**Comment by mcabbott** [comment]
Why not write the modules first, since they are part of the function's full name? Perhaps with colour it may be clear enough just to write `Core.eval` etc, or they could still be inn columns.

**Comment by timholy** [comment]
Really crazy thought: if printing the argument types would take up more than one line, print the signature in folded form and then use `REPL.TerminalMenus` to allow viewers to expand it. Something like:

```
julia> run_command(args...)
    Function    Arguments              Location
[1] f           (x::Int, y::Float64)   MyModule1, /path/to/file1.jl: 15
[2] g          +(...)                  MyModule2, /path/to/file2.jl: 64
...

julia> lasttrace()
# presents the above in menu form, allowing user to expand line 2
```

Related: #35915 (which I plan to use to create a folded-tree menu, https://github.com/JuliaCollections/FoldingTrees.jl).

**Comment by timotaularson** [comment]
How does it look with codepath rearranged in this order: LineNumber Filename Path
Then your eyes have a stable location to look for the line number and filename with no extra color needed for them (I seem to always be searching for the filename and line number buried in the details, can you tell?)
And for interactive REPL work I love the folded idea above :)

**Comment by jkrumbiegel** [comment]
> Why not write the modules first, since they are part of the function's full name

Yeah I kind of like the idea, I think I started out without color and that didn't work so well, but like this it's actually quite readable. I'll try without columns next, although columns are always nice because they guide your eyes.

<img width="603" alt="grafik" src="https://user-images.githubusercontent.com/22495855/82936957-a1af9580-9f8f-11ea-81ca-58407dd3beac.png">

**Comment by JeffBezanson** [comment]
Not sure about that; the function name is a vastly more important piece of information than the module. For me at least, the most important info is definitely the function name, file name, and line number. Also for scripts the left column would be lots of `Main` or blank, which is not ideal for that use case.

**Comment by ericphanson** [comment]
I think when I’m reading stacktraces, I’m mostly looking for the last call in my code before it goes to Base/package code, since usually the bug is in my code. So in that case the module is pretty important and the function call isn’t what I’m scanning first.

**Comment by JeffBezanson** [comment]
How about this:
```
[1]  get_ticklabels   ::AbstractPlotting.Automatic, ::Int64
     in MakieLayout at ~/.julia/packages/MakieLayout/COfBu/src/lineaxis.jl:372
```
Module names would still line up so they would be easy to scan.

**Comment by jkrumbiegel** [comment]
I had actually just tried out something close to that 
<img width="553" alt="grafik" src="https://user-images.githubusercontent.com/22495855/82948784-bf3a2a80-9fa2-11ea-9fc2-0be302cff2e1.png">

**Comment by JeffBezanson** [comment]
The first time I saw that I'd probably go nuts trying to figure out what the colors meant :joy: If there are going to be per-module colors, seems better to just apply the color to the module name.

**Comment by mcabbott** [comment]
I like the module names, they get a bit lost in this latest picture. 

If they are at the bottom, perhaps making them the same bright colour would connect things? (And make clear where you cross a module boundary.) And, perhaps, if the module name appears in the path, it could simply be highlighted there instead -- that would also move the colourful name off the left edge where the functions are.

**Comment by KristofferC** [comment]
For some inspiration, I post a few images from the last time tweaking the stacktrace formatting came up.

![bild](https://user-images.githubusercontent.com/1282691/82949976-d2e69080-9fa4-11ea-9343-515dd8bf610c.png)

![bild](https://user-images.githubusercontent.com/1282691/82950030-ebef4180-9fa4-11ea-8f65-f384d3df478c.png)

**Comment by antoine-levitt** [comment]
Often the exact package is not so much of interest as the nature of the code: is it core julia, a library I'm using, the library I'm developing, or a script? We could have four colors: core/base/dev/stdlib, packages, packages currently `]dev`ed, and all the rest (including script/REPL). Some of these distinctions are pretty arbitrary and the classification is a bit brittle, but coloring the method based on this would (at least for me) maximize the information without displaying any extra text and keeping a small, easy to memorize set of colors.

**Comment by jkrumbiegel** [comment]
For me, the colors are more about boundary changes, where code from one module begins and the other ends. That's why I probably wouldn't assign inherent meaning to them. On the other hand, that might be confusing.

I've tried to simplify more, and removed columns again because without the module column they are not so useful anymore. Then first I colored only line number by modules, but this still left the file name not so visible, which many comments said is important. So I colored that in the module color as well. The module name itself is not colored because it's too close to the function name and that's too noisy.

Here is the version with both numbers and filenames colored:

<img width="559" alt="grafik" src="https://user-images.githubusercontent.com/22495855/82990295-48ca1680-9ffc-11ea-8a50-13152fd6aeb9.png">

**Comment by jkrumbiegel** [comment]
here without filenames colored

<img width="566" alt="grafik" src="https://user-images.githubusercontent.com/22495855/82990461-87f86780-9ffc-11ea-8af3-d7edef156102.png">

**Comment by kmsquire** [comment]
Can you also try coloring the full file path?

**Comment by timotaularson** [comment]
Some ideas...
Filename and line number in a consistent location with double spaces to set them off.
Stack level color and module color match each other to visually pair the related lines.
Filename, line number, and path in slightly different hue than parameter types to distinguish without clutter.
![stacktrace](https://user-images.githubusercontent.com/44177396/83088654-574b1900-a062-11ea-859a-d43b3c0f5651.png)
[Colors in this example are arbitrary, more thinking about how their positions relate.
Missing stack levels in example and shortened parameter type list on last item are because I typed example by hand.]

**Comment by timotaularson** [comment]
Code used to produce the sample above in case anyone wants to play with it:
```
function main()

    errors = [
              ("1", "get_ticklabels", ("AbstractPlotting.Automatic", "Int64"), "372", "lineaxis.jl", "MakieLayout", "~/.julia/packages/MakieLayout/COfBu/src")
              ("2", "get_ticklabels", ("AbstractPlotting.Automatic", "Int64"), "351", "lineaxis.jl", "MakieLayout", "~/.julia/packages/MakieLayout/COfBu/src")
              ("3", "#105", ("AbstractPlotting.Automatic",), "152", "lineaxis.jl", "MakieLayout", "~/.julia/packages/MakieLayout/COfBu/src")
              ("8", "OnUpdate", ("Tuple{Float32,Float32}",), "218", "Observables.jl", "Observables", "~/.julia/packages/Observables/0wrF6/src")
              ("9", "#setindex!#5", ("Observables.var\"#6#8\"",), "138", "Observables.jl", "Observables", "~/.julia/packages/Observables/0wrF6/src")
              ("10", "setindex!", ("Observables.Observable{Any}",), "126", "Observables.jl", "Observables", "~/.julia/packages/Observables/0wrF6/src")
              ("11", "#LineAxis#95", ("Base.Iterators.Pairs{Symbol,Observables.Observable,NTuple{28,Symbol},NamedTuple{(:endpoints, :limits, :flipped, :ticklabelrotation, :ticklabelalign, :lables
ize, :labelpadding, :ticklabelpad, :labelvisible, :label, :labelfont, :ticklabelfont, :ticklabelcolor}",), "270", "lineaxis.jl", "MakieLayout", "~/.julia/packages/MakieLayout/COfBu/src/lobjects")
    ]

    println()
    for (idx, err) in enumerate(errors)
        # Module color
        mc = idx <= 3 ? :light_red : (idx >= 7 ? :light_red : :light_yellow)
        # Path color
        pc = :blue
        printstyled("[", color=mc)
        printstyled("$(err[1])", color=mc)     # errorno
        printstyled("]  ", color=mc)
        printstyled("$(err[5])", color=pc)     # filename
        printstyled(":", color=pc)             # colon
        printstyled("$(err[4])", color=pc)     # lineno
        printstyled("  $(err[7])", color=pc)   # path
        println()
        printstyled("$(err[6]) ", color=mc)    # module
        printstyled("$(err[2]) ", color=:bold) # function
        printstyled("(", color=:light_blue)    # param types
        for t in err[3]
            printstyled("::", color=:white)
            printstyled("$t", color=:light_blue)
        end
        printstyled(")", color=:light_blue)
        println()
    end
end

```

**Comment by jkrumbiegel** [comment]
The problem with this is that this doesn't leave the links clickable in Atom / VSCode, etc. This is something I use very often and I believe others do as well. In fact, I even explicitly expand the base paths to make them clickable. This does of course print more. But I think it increases utility. I know there's some way to jump to stack trace entries by number in the REPL with some shortcut, but this is far less intuitive than just clicking on a link in my opinion.

**Comment by KristofferC** [comment]
It might make sense to take the opportunity to unify the way line info is printed with the logging system, for example:

![bild](https://user-images.githubusercontent.com/1282691/83130387-869f6b80-a0de-11ea-82f8-601844a7ce13.png)

**Comment by KristofferC** [comment]
Slightly more conservative attempt:

![bild](https://user-images.githubusercontent.com/1282691/83138583-4eeaf080-a0eb-11ea-9db5-24de295512f3.png)

**Comment by mcabbott** [comment]
A version with variable names added in, and colour just on the module names:

<img width="736" alt="Screenshot 2020-05-28 at 15 55 56" src="https://user-images.githubusercontent.com/32575566/83150578-ba3cbe80-a0fb-11ea-9e68-6d38053e4cda.png">

This one doesn't have very long type signatures, but perhaps having variable names like this (if that's possible?) will make it easier to spot the outermost type of each argument. 

<details><summary>Adapting @timotaularson 's code</summary>
<pre>
let
    printstyled("\njulia> ", color=:green)
    println("f()")
	printstyled("""ERROR: DimensionMismatch("A has dimensions (1,2) but B has dimensions (3,4)")""", color=:light_red)
	println("\nStacktrace:")
    for (idx, err) in enumerate(errors)
        mc = idx <= 3 ? :blue : (idx >= 7 ? :blue : :yellow) # Module color
        printstyled("[$(err[1])] ", color=:normal, bold=true) # errorno
        printstyled(err[2], color=:normal, bold=true) # function
        printstyled("(", color=:normal)    # param types
        for (i,t) in enumerate(err[3])
        	i != 1 && printstyled(", ", color=:normal)
        	printstyled(rand('a':'z')^2, color=:light_black)
            printstyled("::", color=:normal)
            printstyled("$t", color=:normal)
        end
        printstyled(")", color=:normal)
        println()
        printstyled("  @ ", color=:light_black)
        printstyled(err[6], color=mc)    # module
        printstyled(" $(err[7])/$(err[5]):$(err[4])", color=:light_black)   # path
        println()
    end
end
</pre>
</details>

**Comment by jkrumbiegel** [comment]
> I think there are two core ideas in the proposal here:
> 
>     * Have the line info on a separate line from the signature (and perhaps delimited it with some color to make it easier to find).
> 
>     * Show the module.
> 
> 
> So here is a slightly more conservative attempt with both those ideas:

I like this. I think the term is "kill your darlings", and my darling might be the color. But I can see how something toned-down might be better suited for something so basic. I changed it a bit so that the module name and the function name line up, which creates an almost column-like vertical contrast line. That should help to find the relevant information. Though I still think I wouldn't put the types in white, because they get so messy. If we had variable names, maybe those in white would look good.

<img width="584" alt="grafik" src="https://user-images.githubusercontent.com/22495855/83157022-6afa8c00-a103-11ea-810e-1e3d9640146f.png">

**Comment by KristofferC** [comment]
> I like this. I think the term is "kill your darlings", and my darling might be the color. 

Just so you know that I know your situation, https://github.com/JuliaLang/julia/pull/18228. This is a topic that is very fun to bikeshed and everyone has a bucket of paint, heh. As you can see, that PR also started quite ambitiously but slowly got more and more conservative heh. 

To me, the last proposal is definitely an improvement over status quo. Two things I am thinking:

- If possible, it would be nice to have some kind of "hook" for your eyes for the line information because it tend to blend together with the signature when the signature is long.
- We should probably align the stackframe numbers when there are more than 10 of them:
    ```
    [ 8]
    [ 9]
    [10]
    [11]
    ```
    etc. That would make the `@` also line up.

**Comment by JeffBezanson** [comment]
I like it, I think we're getting somewhere. Being judicious with colors is always good, since that exposes less surface area for bikeshedding, plus decreases the risk that a color will be unreadable on somebody's terminal.

Spacing things out is good, but unfortunately the syntax `f (...)` is explicitly disallowed, so I think we need to either remove the space or remove the parens.

**Comment by timotaularson** [comment]
The most recent sample generally looks good, but I second the idea:

> If possible, it would be nice to have some kind of "hook" for your eyes for the line information because it tend to blend together with the signature when the signature is long.

And I hope that hook is a different color or (preferably) brightness.

**Comment by timotaularson** [comment]
Maybe you could do the cycle of colors for the modules but only apply it to the [ ] on either side of the stack level number so if you cannot see the colors for some reason not much is lost.

**Comment by jkrumbiegel** [comment]
> Maybe you could do the cycle of colors for the modules but only apply it to the [ ]

It's hard to even see the color correctly if it's only brackets and they are in a contrasty area, so I don't think that works too well..

> the syntax f (...) is explicitly disallowed

I think it works well even without the space, because of the contrast difference.

> If possible, it would be nice to have some kind of "hook" for your eyes for the line information 

I agree, I think it works well to leave the line numbers gray but bold them, that makes them just slightly salient if you look for them but the next step of making them white is way too noisy in my opinion.

> We should probably align the stackframe numbers 

I tried it with spaces inside the brackets like in your example but found that looked a bit weird, so now I'm just right aligning the bracketed numbers. I think that works pretty well.

Here's the newest version with all those ideas incorporated:

<img width="578" alt="grafik" src="https://user-images.githubusercontent.com/22495855/83169433-51614080-a113-11ea-81bc-1f94551a7d58.png">

**Comment by jkrumbiegel** [comment]
And as a bonus, here's how it could look with (random) variable names. Again I'm just using bold gray because all this variable names should not fight for your attention, but should be a bit discernible if you look for them 
<img width="578" alt="grafik" src="https://user-images.githubusercontent.com/22495855/83171530-8cb13e80-a116-11ea-9092-14ba32cbc2cb.png">

**Comment by StefanKarpinski** [comment]
I do like where this is going too. I want to just put a vote in for some kind of highlighting of the module name. I actually quite like the style where the same module is colored the same way and it cycles through the colors (there's a graph color problem in here if you consider two stack trace lines being adjacent as a graph edge between their modules). It is, unfortunately, pretty non-obvious what that color mean, however. Perhaps coloring all the module names in a single color?

**Comment by mcabbott** [comment]
This alignment looks good to me. 

I still think that having the paths more clearly distinct from the signatures would help break things up -- 3 levels of importance, name/signature/path. Logging macros `@info` etc. print them ligher than other lines.

Variable names would be extremely nice to have. Would vote that they be less significant than signature information, probably?

Having the module names lined up at the left does help you scan them, although I'm sad to see colours go.

**Comment by timotaularson** [comment]
My eyes seem to know where to find things on this one :)
Would a blank line between modules use too much vertical space?

**Comment by ViralBShah** [comment]
Coloring rows with 2 different colors so that alternate rows have the same color - as is often done in tables might help readability.

**Comment by timotaularson** [comment]
Color cycling the @ next to the module name instead of the module name itself stands out (because @ is a pretty big character) without competing with the highlighted function name too much.

I have often wished for the parameter names since they help you find the right parameter type without having to count through them.

**Comment by mcabbott** [comment]
You could explicitly mirror what logging uses, although ideally with different colours:
```julia
let
    printstyled("┌[", color=:magenta, bold=true); print("1"); printstyled("] ", color=:magenta, bold=true)
    printstyled("get_ticklabels", bold=true); print("("); printstyled("style", color=:light_black); print("::AbstractPlotting.Automatic, "); printstyled("n", color=:light_black); println("::Int64)")
    printstyled("└ @ ", color=:magenta, bold=true); printstyled("MakieLayout", color=:magenta); printstyled(" ~/.julia/packages/MakieLayout/COfBu/src/lineaxis.jl:372", color=:light_black); println("\n")
end
@error "compare" pi
```

**Comment by jkrumbiegel** [comment]
The thing that I don't like so much about this is that line breaks that are only there because of the specific REPL width don't transfer well if you resize your window / REPL (if you want to make room for longer lines for example) or copy / paste the text into windows of different width

**Comment by mcabbott** [comment]
Yes, I would not propose explicitly breaking the lines, and indeed it may look awful with multi-line expressions. Just a thought on how to tie the colour to the function name, and how to make it perhaps less obtrusive, if it's part of the border instead of looking like the most important heading.

**Comment by karajan9** [comment]
I must say I liked the colour but I can understand the preference to be conservative with it.

I want, however, to throw my support behind something like
> Often the exact package is not so much of interest as the nature of the code: is it core julia, a library I'm using, the library I'm developing, or a script? We could have four colors: core/base/dev/stdlib, packages, packages currently `]dev`ed, and all the rest (including script/REPL). [...]

Since I'm not hacking on base and rarely on packages (as most people do, I presume), 90% of errors need to be fixed in my script, 9% in a package and 1% in base. If I made a mistake calling a function I don't need to know here exactly in base the error was thrown since I'm not going in there to fix it anyway, I need to fix my code.

So I'd appreciate some visual helper to see where I'm leaving my code (and maybe also where I'm leaving package code, going into base) because in most cases anything after that I can safely ignore for now. If not via colour, maybe with `--------------------------------------------` or something?

**Comment by mcabbott** [comment]
If modules get coloured, then perhaps Base/Core should be left out and given no colour.

**Comment by JeffBezanson** [comment]
> I actually quite like the style where the same module is colored the same way and it cycles through the colors

I'm fine with that too. If we're going to use many colors, I think they should be used for that. It also helps pick out boundaries between user and package code without taking up too much vertical space.


> If modules get coloured, then perhaps Base/Core should be left out and given no colour.

Yes, Base and Core could just always be dark gray or something, leaving more colors for everybody else :)

**Comment by mcabbott** [comment]
If this gives up working perfectly in 16 colours, then it might be nice to make them a hash of the module names, so that you will memorise some of them.

**Comment by KristofferC** [comment]
Or only have two colors, one for Base/Core and one for external modules?

**Comment by jkrumbiegel** [comment]
Originally, I thought it shouldn't be much of a problem to cycle colors because my color scheme has at least 6:
```julia
crayon"blue",
crayon"yellow",
crayon"red",
crayon"green",
crayon"cyan",
crayon"magenta",
```

There are light variants too I guess, but my color scheme has them set the same as the others, so I assume that could happen sometimes. But how likely is it that a stack trace goes through 6 different non-base or standard library modules? Of course packages use code from many modules, but in one particular stack there shouldn't be *that* many.

**Comment by JeffBezanson** [comment]
> then it might be nice to make them a hash of the module names, so that you will memorise some of them

This is actually kind of brilliant.

**Comment by laborg** [comment]
Great to see people working on this. 
I'd be interested to see how right alignment of the function names looks:
```
 [1] get_ticklabels(code_lowered::...
 [2] get_ticklabels(real::AbstractPlotting
 [3] #105(mtime::
 [4] OnUpdate(vecormat::
 [5] #setindex!#5(atexti::
```
vs
```
 [1] get_ticklabels(code_lowered::...
 [2] get_ticklabels(real::AbstractPlotting
 [3]           #105(mtime::
 [4]       OnUpdate(vecormat::
 [5]   #setindex!#5(atexti::
 [6]      setindex!(mapslices
```
It might be easier for the eye to catch where the function arguments start and offer assistance on parsing the chain of function calls while not wasting much space. It still needs to be supported by colors, otherwise the locations infos and arguments would drown everything.

**Comment by karajan9** [comment]
> > then it might be nice to make them a hash of the module names, so that you will memorise some of them
> 
> This is actually kind of brilliant.

Oh, hello darkish teal, my old friend...

**Comment by antoine-levitt** [comment]
> then it might be nice to make them a hash of the module names, so that you will memorise some of them

I used to have emacs do exactly that for IRC nicknames. It didn't work out so well because some of my friends got assigned the same color and it was more confusing than anything, so I ended up hard-coding colors for my friends. I wouldn't recommend going down that route, however cool it seems. Having a reduced set of colors for each code type (eg Julia, package, deved package, user) seems more useful.

**Comment by mcabbott** [comment]
I landed on an editor colour scheme which does this, and it's moderately useful -- all the variables which might otherwise be blue are various shades of green & blue, with strings & keywords etc. different. Perhaps the ideal would be something like within broad classes like the ones you suggest (e.g. standard library in red-purple, downloaded in green-blue,  dev-ed in orange-yellow, base in gray). Or perhaps this is much too much work!

**Comment by jkrumbiegel** [comment]
another iteration: I tried getting variable names in there, even though it all feels super hacky to me as I'm not very familiar with all the internals. I re-colored the modules because it seemed to get a positive response in general, even though it's debatable how the colors should be assigned. But with the colors, I feel that the module of one entry clashes a bit with the function name of the next. So I introduced line breaks. Makes everything longer of course, but I kind of like the additional breathing room and clarity 
<img width="627" alt="grafik" src="https://user-images.githubusercontent.com/22495855/83299019-bb521680-a1f5-11ea-9927-fa3f3064987c.png">

**Comment by timotaularson** [comment]
![stacktrace2](https://user-images.githubusercontent.com/44177396/83304530-06a3ff00-a1cd-11ea-9b40-d14d57d9a649.png)

**Comment by JeffBezanson** [comment]
@jkrumbiegel That's beautiful. I need this in my life.

**Comment by vtjnash** [comment]
I like those soft blues

**Comment by smldis** [comment]
@jkrumbiegel Great, I am not sure the @ symbol is needed, it might look cleaner without it

**Comment by el-oso** [comment]
@jkrumbiegel That looks great! Is there a way to highlight just the filename (not the full path) and line number? Perhaps a slightly brighter grey?

**Comment by StefanKarpinski** [comment]
Lovely! I actually really like the line breaks—they really help visually and our stack traces tend to be so long I have to scroll anyway, I’d rather it be clear than reduce the scrolling by a little bit.

**Comment by karlwessel** [comment]
The highlighted `::` are brilliant. If I would have to plot a curve showing the information content of each character of the parameter list against character position, the `::` exactly mark the local maxima of that curve. I know to look to left of them for the name and to the right for the most important part of the type, after that the information content only drops.

**Comment by antoine-levitt** [comment]
I disagree with this comment: the highlighting should emphasize important things so that the eye knows what to read, and the :: definitely is not an important thing I should read. As it stands the white is much more noticeable than the bold dark blue. Distinguishing the parameters would be clearer by just having in white the argument names (perhaps with the exception of when there's no named argument, in which case `::` might as well be in white to delimit arguments). Same goes for the line number, I'd rather emphasize the file name than the line number. 

The colored package names run the risk of overshadowing the most important thing, the function name, but having those in bold and with linebreaks really limits that risk, well done!

**Comment by jkrumbiegel** [comment]
I thought about the :: a lot before I made them white. Of course the argument names and the types are the actual information, so you might think they should be highlighted. But my strong impression when highlighting them is that they call for your attention. But it’s not important to read every name and type. You should just be able to find those that you’re looking for in a specific function. How I see the steps of reading a stack trace:

1. Get a general overview
2. What function in what module errored?
3. What other functions called this function from which other modules?
4. Is there a critical boundary between my own code and base or other package code? Where is it?
5. What function is at that boundary?
6. What types / arguments were used?
7. Does that explain the error?
8. If this didn’t help, go through everything with a fine comb

So the type / argument information is in my opinion used only after orienting and understanding the general structure. One thing to understand about highlighting is that it doesn’t necessarily make it easier to find things, only if the highlights are relatively sparse. So I think the :: highlight doesn’t get in the way and grab your attention until you start looking for variable / type information. And at that step it gives your eyes hooks to jump to, nothing more, nothing less. I find that quite efficient.

As a side node, not every argument has a name, so highlighting variable names would not always give you good separation. I think the :: is always there.

**Comment by mcabbott** [comment]
That seems like a good list, my main complaint is point 6, finding the gaps in long type signatures. Printing `::` super-bright does help, although I agree it's a little weird. Printing the argument names (or perhaps some `_1` placeholder if none) would help a bit. Perhaps printing the outermost type normally, and then all its arguments in gray, would help a lot? 

How many levels of highlighting are actually available? `:white` won't stand out on light backgrounds, in fact I think we may be limited to normal, bold, and `:light_black` (plus colours). Right now the function name and the path are bold. In `@error` etc, the path is light_black, which seems great.

I like the use of colour to highlight modules. In some of the samples above it outshines the other text, but is this true of simple colours? In a few different themes, "Module" here is relatively muted (and `::Array, ::Adjoint, ::Int` pretty clearly the top-level signature):
```
let
    printstyled("\nfunction", bold=true); print("(x::Array"); printstyled("{Float64,1}", color=:light_black); print(", y::Adjoint"); printstyled("{Float64,2,Array{Float64,2}}", color=:light_black); println(", z::Int64)")
    printstyled(" @ ", color=:light_black); printstyled("Module", color=:blue); printstyled(" ~/.julia/dev/Package/src/Package.jl:33 \n", color=:light_black)
end
```
<img width="733" alt="Screenshot 2020-05-31 at 16 14 35" src="https://user-images.githubusercontent.com/32575566/83354599-ff5e2c00-a359-11ea-9cca-0cc5cb374d40.png">

**Comment by singularitti** [comment]
Might be a bit off-topic but since we have 

> another iteration: I tried getting variable names in there, even though it all feels super hacky to me as I'm not very familiar with all the internals. I re-colored the modules because it seemed to get a positive response in general, even though it's debatable how the colors should be assigned. But with the colors, I feel that the module of one entry clashes a bit with the function name of the next. So I introduced line breaks. Makes everything longer of course, but I kind of like the additional breathing room and clarity
> 
> <img alt="grafik" width="633" src="https://user-images.githubusercontent.com/22495855/83299525-a2963080-a1f6-11ea-86e3-7f1aeb77b9a6.png">

Since we have this, can we have a prettier `show` for the output of `methods`? It is also a pain.

**Comment by singularitti** [comment]
> @jkrumbiegel That looks great! Is there a way to highlight just the filename (not the full path) and line number? Perhaps a slightly brighter grey?

It could: https://github.com/JuliaLang/julia/issues/36026#issuecomment-634481656. But it is a bit weird to me since the file name is part of the path so why it uses a different color? Usually, I just click the path and VSCode will open it for me so it doesn't matter what name the file is.

**Comment by jkrumbiegel** [comment]
I've moved away from using Crayon.jl, now that I don't actually use complex colors anymore. I found a way to enhance the visibility of file name and line number without it becoming visually overwhelming by underlining in dark grey. That looks sensible, too, as we're used to paths / links being underlined. White or other highlighting was too strong there, bold too weak.

Also, I've switched to light colors as the first colors in the cycler, which I should have done in the first place, but they look the same in my theme so I never noticed. This should be better for themes where dark blue is hardly visible (that's the theme's fault though).

I've registered this style as version 0.2 in `ClearStacktrace.jl` so we can try it out a bit more.

Two examples:

<img width="577" alt="example 1" src="https://user-images.githubusercontent.com/22495855/83422129-5c74e300-a429-11ea-9bc4-6e765f04c490.png">

<img width="579" alt="example 2" src="https://user-images.githubusercontent.com/22495855/83422813-5df2db00-a42a-11ea-80c5-4a7316659cc2.png">

**Comment by timotaularson** [comment]
That is really nice work.
You have the parameter names lighter and the types darker. Did that turn out to look better/read better that way than with the types lighter and the names darker?

**Comment by jkrumbiegel** [comment]
Yes, as the types are usually much longer, it doesn’t help much to make them lighter.

**Comment by StefanKarpinski** [comment]
Random thoughts:

* maybe horizontal rules between groups of stack trace entries in the same module? maybe too busy looking
* might be good at the top of the stack trace too, to separate it clearly from the error message

**Comment by KristofferC** [comment]
> I've moved away from using Crayon.jl, now that I don't actually use complex colors anymore. Helps with the load time of the package, too.

Did it cause a significant load time? It loads pretty quickly on its own for me

```
julia> @time using Crayons
  0.014410 seconds (22.60 k allocations: 2.274 MiB)
```

**Comment by JeffBezanson** [comment]
This version is great, let's put this in Base.

**Comment by StefanKarpinski** [comment]
Agree: let’s just do that last version.

**Comment by JeffBezanson** [comment]
Note in the current stacktrace printing frames 8-11 in that example are not shown (they are part of the REPL and would be in every REPL stacktrace).

**Comment by mcabbott** [comment]
Indeed, this has improved which is great. But it still goes from 8 lines (for the stack trace alone) to 20 (ClearStacktrace 0.2). There's a bit of a trade-off between how pretty it is & not losing your place.

**Comment by jkrumbiegel** [comment]
The base paths are expanded on purpose to make them clickable. You can disable that in ClearStacktrace. I could also make the newlines optional for people who don’t like them. Could just be an environment variable.

**Comment by jkrumbiegel** [comment]
And I guess I must have missed copying the part of the function that clips off the last couple of frames that never change

**Comment by singularitti** [comment]
> I've moved away from using Crayon.jl, now that I don't actually use complex colors anymore. Helps with the load time of the package, too. I found a way to enhance the visibility of file name and line number without it becoming visually overwhelming by underlining in dark grey. That looks sensible, too, as we're used to paths / links being underlined. White or other highlighting was too strong there, bold too weak.
> 
> Also, I've switched to light colors as the first colors in the cycler, which I should have done in the first place, but they look the same in my theme so I never noticed. This should be better for themes where dark blue is hardly visible (that's the theme's fault though).
> 
> I've registered this style as version 0.2 in `ClearStacktrace.jl` so we can try it out a bit more.
> 
> Two examples:
> 
> <img alt="example 1" width="577" src="https://user-images.githubusercontent.com/22495855/83422129-5c74e300-a429-11ea-9bc4-6e765f04c490.png">
> 
> <img alt="example 2" width="579" src="https://user-images.githubusercontent.com/22495855/83422813-5df2db00-a42a-11ea-80c5-4a7316659cc2.png">

Curious about why there is an extra `/` in `//Applications/Julia-1.4.app/`?

**Comment by jkrumbiegel** [comment]
Probably a bug from splitting and rejoining the path

**Comment by jkrumbiegel** [comment]
> > I've moved away from using Crayon.jl, now that I don't actually use complex colors anymore. Helps with the load time of the package, too.
> 
> 
> 
> Did it cause a significant load time? It loads pretty quickly on its own for me
> 
> 
> 
> ```
> 
> julia> @time using Crayons
> 
>   0.014410 seconds (22.60 k allocations: 2.274 MiB)
> 
> ```

No, I mostly removed it because I wouldn’t have it in base :) Load time was just a guess

**Comment by singularitti** [comment]
I guess there might be too many blank lines if we have many frames of stacktrace as shown in https://github.com/JuliaLang/julia/issues/36026#issuecomment-636912686 ?

**Comment by KristofferC** [comment]
Not to derail the discussion (the last version is great and a big improvement) but on the topic of too many newlines, the problem is that when working at the terminal it seems a lot better to me to print the stacktrace "inverted" (as I originally proposed in https://github.com/JuliaLang/julia/pull/18228) as:

```
...
[3] frame
[2] frame
[1] frame
Error: Some error happened
blah blah
```

The most important information is the error message itself and the frames towards the top of the stack (closer to the error) and printing it in this order that will always be visible. Right now I frequently find myself having to scroll up to even see the error message and not just the tail of a stacktrace.

However, when reading a stacktrace on a webstite that has been copy-pasted from the terminal you want the opposite order because you scroll from top to bottom as opposed to bottom to top as you do in a terminal... Kinda tricky to get a good design for both these cases.

**Comment by jkrumbiegel** [comment]
> However, when reading a stacktrace on a webstite that has been copy-pasted from the terminal you want the opposite order because you scroll from top to bottom as opposed to bottom to top as you do in a terminal... Kinda tricky to get a good design for both these cases.

I actually had code in `ClearStacktrace.jl` for a moment that allowed me to reprint the last stacktrace. I had envisioned that for cutting off super long types at some character maximum and being able to reprint in full if the whole information was needed. But your use case would also be interesting. I can imagine a `reprint(inverted = true)` or `reprint(html = true)` where it would print a html version that would retain coloring for pasting into a website.

Also, I agree that given the scrolling direction it could make sense to invert the whole thing by default.

**Comment by c42f** [comment]
`ipython` prints frames in the opposite order, and despite not having to scroll I've always found it inexplicably confusing. Perhaps that's just due to my prior experience with gdb and other systems where the innermost frame is at the top, or perhaps other people feel this way too?

Speaking of gdb, they have a reasonable solution to very long traces with their pager: "press enter for more frames".

**Comment by c42f** [comment]
By the way, I Iove the latest design in https://github.com/JuliaLang/julia/issues/36026#issuecomment-636912686 and would be extremely happy if we merge that. Changing the frame order or adding interactive features seem like separate issues.

Regarding file names, I hope that we can eventually use terminal hyperlink OSC sequences ([yes hyperlinks in terminals are somewhat widely supported!](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda)) and have a way for the user's editor to pick that up.

**Comment by ggggggggg** [comment]
Speaking of finding the "innermost frame", I switch languages enough that I can never remember if I should be looking at the top or the bottom for my code. So I end up scanning through looking at the filename until I see one I recognize. The underlining shown here would help, so this is a clear improvement. But I still wonder if there is a goodway to call out if I should be looking at the top or the bottom. In principle printing `YOUR CODE HERE` on one end and `OTHER CODE HERE` on the other end would help me, but it doesn't seem very elegant.

**Comment by jkrumbiegel** [comment]
I've made a PR to Base here https://github.com/JuliaLang/julia/pull/36134
It works as far as I can tell, but I'll need some help to make it ready for merging

**Comment by ChrisRackauckas** [comment]
Is there a way to omit type parameters? The problem we see a lot of the time is that the amount of type parameters in DiffEq, ForwardDiff, etc. can make things... daunting. If by default it just said `Dual`, except in the case where there is dispatch to other methods due to type parameters, then I think you'd reduce 90% of the noise in most stack traces I read (and in the other cases, @timholy 's suggestion is really what's required, since it's usually about making types or matching type parameters)

**Comment by vtjnash** [comment]
If they correspond to some existing type alias, they'll just go away automatically now (#36107). Otherwise, they indicate possible compilation performance bottlenecks—so possibly worth investigating?

**Comment by KristofferC** [comment]
This is done now.

**Comment by ChrisRackauckas** [comment]
For some people, but most people just get confused when it prints out a type that would take 3 pages of printed paper, so it should probably be something that's opt-in. I'll open up another issue to discuss that.

**Comment by ChrisRackauckas** [comment]
Continued in https://github.com/JuliaLang/julia/issues/36517
